# Week 2 Checkpoint: First Data Profile
**BAN 6003 Data Management and Analytics Integration**

Complete this notebook using the `employee_data.csv` file in the `data/` folder.

This is the independent checkpoint for Week 2. Use the functions practiced in the guided lab. You are not expected to clean the data yet. Your goal is to profile the data and report what you notice.


## Instructions

1. Run each cell from top to bottom.
2. Fill in all cells marked `# Your code here`.
3. Write short answers in the markdown response areas.
4. Save, commit, and push your notebook to GitHub.

You may use the Week 2 guided lab as a reference.


## Task 0: Import Pandas and Load the Data
Run the cell below.


In [18]:
import pandas as pd

df = pd.read_csv("../data/employee_data.csv")
df.head()


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
0,101,John,Smith,4/15/22,Marketing,85000,34.0,True
1,102,Jane,Doe,8/20/21,Sales,92000,45.0,True
2,103,Peter,Jones,1/2/22,HR,71000,29.0,True
3,104,Mary,Johnson,11/10/23,Engineering,115000,38.0,True
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False


## Task 1: Basic Structure
Show the basic structure of the DataFrame.

Your code should show:
- the shape of the DataFrame;
- the column names;
- the first five rows;
- the last five rows.


In [70]:
print(df.shape)
print(df.columns)
display(df.head(5))
display(df.tail(5))

print(df)

(13, 8)
Index(['employee_id', 'first_name', 'last_name', 'hire_date', 'department',
       'salary', 'age', 'full_time'],
      dtype='str')


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
0,101,John,Smith,4/15/22,Marketing,85000,34.0,True
1,102,Jane,Doe,8/20/21,Sales,92000,45.0,True
2,103,Peter,Jones,1/2/22,HR,71000,29.0,True
3,104,Mary,Johnson,11/10/23,Engineering,115000,38.0,True
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
8,102,Emily,White,1/15/24,Sales,91000,31.0,True
9,110,David,Miller,7/22/19,HR,68000,58.0,True
10,111,Sarah,Moore,3/12/23,Marketing,"$91,000",28.0,True
11,error,Kevin,Clark,2/5/24,Sales,88000,35.0,True
12,113,Brian,Hall,6/18/22,Engineering,5900000,999.0,True


   employee_id first_name last_name    hire_date   department    salary  \
0          101       John     Smith      4/15/22    Marketing     85000   
1          102       Jane       Doe      8/20/21        Sales     92000   
2          103      Peter     Jones       1/2/22           HR     71000   
3          104       Mary   Johnson     11/10/23  Engineering    115000   
4          105       Mike     Brown      5/30/20        sales     89000   
5          106      Susan     Davis          NaN        Sales     95000   
6          107      Chris    Wilson  2023-FEB-21    Marketing       NaN   
7          108      Laura    Taylor       9/1/21  Engineering    120000   
8          102      Emily     White      1/15/24        Sales     91000   
9          110      David    Miller      7/22/19           HR     68000   
10         111      Sarah     Moore      3/12/23    Marketing  $91,000    
11       error      Kevin     Clark       2/5/24        Sales     88000   
12         113      Brian

### Task 1 Written Response
Write 1–2 sentences describing the size and general structure of the dataset.

**Your response:**  
The data consists of 13 rows and 8 columns. All columns are dtype string and will need to be converted to correct format.


## Task 2: Data Types and Non-Missing Counts
Use Pandas to inspect data types and non-missing counts.

Your code should use:
- `info()`
- `dtypes`


In [26]:
# Your code here
display(df.info())
df.dtypes


<class 'pandas.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   employee_id  13 non-null     str    
 1   first_name   13 non-null     str    
 2   last_name    13 non-null     str    
 3   hire_date    12 non-null     str    
 4   department   13 non-null     str    
 5   salary       12 non-null     str    
 6   age          12 non-null     float64
 7   full_time    12 non-null     object 
dtypes: float64(1), object(1), str(6)
memory usage: 964.0+ bytes


None

employee_id        str
first_name         str
last_name          str
hire_date          str
department         str
salary             str
age            float64
full_time       object
dtype: object

### Task 2 Written Response
Which columns, if any, have data types that might need to be investigated later? For example, is a number or date stored as text/object?

**Your response:**  
Several columns have incorrect data types. EmployeeID and Age should be integer. Salary should be currency or accounting. Hire Date should be formatted as a date.

## Task 3: Missing Values
Check missing values for all columns.

Your code should use `isna().sum()`.

Optional: also calculate missing percentages.


In [38]:
# Your code here
print(df.isna().sum())
print(df.isna().mean() *100)
print(len(df[df.isna().any(axis=1)]))
df[df.isna().any(axis=1)]

employee_id    0
first_name     0
last_name      0
hire_date      1
department     0
salary         1
age            1
full_time      1
dtype: int64
employee_id    0.000000
first_name     0.000000
last_name      0.000000
hire_date      7.692308
department     0.000000
salary         7.692308
age            7.692308
full_time      7.692308
dtype: float64
4


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False
5,106,Susan,Davis,NaN,Sales,95000,41.0,True
6,107,Chris,Wilson,2023-FEB-21,Marketing,NaN,33.0,True
7,108,Laura,Taylor,9/1/21,Engineering,120000,52.0,NaN


### Task 3 Written Response
Which columns have missing values? Which missing values seem most important to investigate later?

**Your response:**  
There are missing values in the hire_date, salary, age, and full_time columns. I think they are all equally important to investigate.


## Task 4: Numerical Summary
Use `.describe()` to summarize numerical columns.

Then choose one numerical column and calculate at least two individual summary statistics such as mean, median, min, max, or standard deviation.


In [54]:
# Your code here
print(df.describe())
agemean = df["age"].mean()
agemin = min(df["age"])
agemax = max(df["age"])
print(f"The mean of age is {agemean}")
print(f"The min of age is {agemin}")
print(f"The max of age is {agemax}")


              age
count   12.000000
mean   118.583333
std    277.411889
min     28.000000
25%     32.500000
50%     36.500000
75%     46.750000
max    999.000000
The mean of age is 118.58333333333333
The min of age is 28.0
The max of age is 999.0


### Task 4 Written Response
What do you notice from the numerical summaries? Mention at least one range, unusual value, or pattern.

**Your response:**  
I notice that there is only one numerical column,age, even though there should be more. Also max age is unrealistic.


## Task 5: Categorical Summary
Use `.value_counts()` to summarize at least one categorical column.

Start with `department` if it exists in the dataset. Then choose one additional categorical column if available.


In [61]:
# Your code here
print(df["department"].value_counts())
print(df["full_time"].value_counts())

department
Sales          4
Marketing      3
Engineering    3
HR             2
sales          1
Name: count, dtype: int64
full_time
True     11
False     1
Name: count, dtype: int64


### Task 5 Written Response
What are the most common categories? Do you notice anything that might need standardization or further investigation later?

**Your response:**  
The most common department is the Sales Department. Most employees are full time. There are 13 rows but only 12 values in the full_time column. That means there are NAs or errors that require investigation.


## Task 6: Duplicate Checks
Check for duplicate records and duplicate employee IDs.

Your code should check:
- fully duplicated rows;
- duplicate `employee_id` values;
- the most common `employee_id` values using `value_counts()`.


In [69]:
# Your code here
full_duplicate_rows = df.duplicated().sum()
print(f"Number of fully duplicated rows: {full_duplicate_rows}")

employee_id_duplicates = df["employee_id"].duplicated().sum()
print(f"Number of duplicated employee_id values after the first occurrence: {employee_id_duplicates}")

print(df["employee_id"].value_counts())

print(df[df["employee_id"] == "102"])

Number of fully duplicated rows: 0
Number of duplicated employee_id values after the first occurrence: 1
employee_id
102      2
101      1
103      1
104      1
105      1
106      1
107      1
108      1
110      1
111      1
error    1
113      1
Name: count, dtype: int64
  employee_id first_name last_name hire_date department salary   age full_time
1         102       Jane       Doe   8/20/21      Sales  92000  45.0      True
8         102      Emily     White   1/15/24      Sales  91000  31.0      True


### Task 6 Written Response
Are there duplicate employee IDs? If yes, what might that mean if `employee_id` is supposed to uniquely identify each employee?

**Your response:**  
Employee 102 appears in 2 rows. It's possible that employee has multiple roles in the company or switched from part-time to full-time. Maybe that employee left and came back. Upon further investigation I see that that ID is associated with 2 different employee names. This needs to be corrected.


## Task 7: Overall Data Quality Observations
Write 3–5 bullet points summarizing the condition of the dataset.

Focus on what you found from profiling. Do **not** clean the data yet.

Examples:
- The dataset has 13 rows and 8 columns.
- The salary column has missing values.
- The hire_date column may have an incorrect data type.
- The employee_id column has possible duplicate IDs.
- The age column has a range from 28 to 999.


**Your observations:**

1. We need to correct data types on id, hire date, salary, age columns
2. Max age is 999. That is bad data.
3. Employee id 102 is a duplicate
4. Department values need to be consistent. Sales vs sales.
5. The data needs to be cleaned before analysis.


## Submission Checklist
Before submitting, make sure:

- [ ] All code cells run without errors.
- [ ] You completed all tasks.
- [ ] You wrote short responses in the markdown cells.
- [ ] You saved the notebook.
- [ ] You committed and pushed your work to GitHub.

Use the terminal:

```bash
git add .
git commit -m "Completed Week 2 data profile checkpoint"
git push
```
